File: backend/app/utils/pdf_filename.py
from __future__ import annotations

from datetime import date


def _financial_year_label(bill_date: date) -> str:
    start_year = bill_date.year if bill_date.month >= 4 else bill_date.year - 1
    return f'{start_year}-{(start_year + 1) % 100:02d}'


def _extract_document_number(document_number: str) -> str:
    candidate = (document_number or '').strip()
    if '/' in candidate:
        candidate = candidate.rsplit('/', 1)[-1].strip()

    sanitized = ''.join(char for char in candidate if char.isalnum() or char in {'-', '_'})
    return sanitized or '0'


def _sanitize_client_name(client_name: str | None) -> str:
    source = (client_name or '').strip()
    sanitized = ''.join(char for char in source if char.isalnum() or char in {'-', '_'})
    return sanitized or 'Client'


def build_bill_pdf_filename(*, bill_date: date, document_number: str, client_name: str | None) -> str:
    year_label = _financial_year_label(bill_date)
    number_label = _extract_document_number(document_number)
    client_label = _sanitize_client_name(client_name)
    return f'{year_label}_{number_label}-{client_label}.pdf'

File: frontend/lib/pdf-filename.ts
const ISO_DATE_PATTERN = /^(\d{4})-(\d{2})-(\d{2})$/;

function financialYearLabelFromIsoDate(isoDate: string): string {
  const match = ISO_DATE_PATTERN.exec(isoDate.trim());
  if (!match) return "0000-00";

  const year = Number(match[1]);
  const month = Number(match[2]);
  if (!Number.isInteger(year) || !Number.isInteger(month) || month < 1 || month > 12) {
    return "0000-00";
  }

  const startYear = month >= 4 ? year : year - 1;
  const endYearTwoDigits = String((startYear + 1) % 100).padStart(2, "0");
  return `${startYear}-${endYearTwoDigits}`;
}

function extractDocumentNumber(value: string): string {
  const trimmed = value.trim();
  if (!trimmed) return "0";

  const maybeSuffix = trimmed.includes("/") ? (trimmed.split("/").pop() || "").trim() : trimmed;
  const sanitized = maybeSuffix.replace(/[^A-Za-z0-9_-]/g, "");
  return sanitized || "0";
}

function sanitizeClientName(value: string | null | undefined): string {
  const source = (value || "").trim();
  if (!source) return "Client";

  const sanitized = source.replace(/[^A-Za-z0-9_-]/g, "");
  return sanitized || "Client";
}

export function buildBillPdfFilename(params: {
  billDateIso: string;
  documentNumber: string;
  clientName: string | null | undefined;
}): string {
  const yearLabel = financialYearLabelFromIsoDate(params.billDateIso);
  const number = extractDocumentNumber(params.documentNumber);
  const client = sanitizeClientName(params.clientName);
  return `${yearLabel}_${number}-${client}.pdf`;
}

